### ISSUE TITLE
Not working on android 10+

### ISSUE THREAD

**Issue by godverrdi** [Magisk-Modules-Repo/Systemless_TZData#25]
Hi
Iran doesn't use daylight savings time anymore but android 10 automatically uses DST 
I don't know how to disable it
I tried disabling Automatic time zone settings and set it manually but some of my apps lost connection with internet 
I was hoping your module would help me but it didn't
Could you please release an update with disabled DST for Iran 
It should be +3:30 but right now it is +4:30

Also tried to replace tzdata from a device thet shows time correctly but didn't work :(

**Comment by kingofnull** [comment]
Try this `tzdata-2022g` files they should cover new DST policy in Iran (all 2021a and newer cover). replace them `tzdata` file with same file in module.

And if you are using miui it may happened because changes doesn't applied correctly. I believe miui doesn't read timezone configuration from `/system/usr/share/zoneinfo/tzdata`. I applied some patches from recovery and when I booted to system and check changed files from android filed explorer they are not there! But if I go back and check from recovery they are applied! (I used OrangeFox recovery to replace files). 

[tzdata-2022g.zip](https://github.com/Magisk-Modules-Repo/Systemless_TZData/files/11051129/tzdata-2022g.zip)

Main generator and files are available from here: https://android.googlesource.com/platform/system/timezone/+archive/refs/heads/pie-dev.tar.gz

**Comment by godverrdi** [comment]
I replaced files and installed module and I can tell that tzdata file has changed successfully but DST is still applied in my region which it should not be.
And I'm using Lineage OS not miui. 
It's not just me... some other android devices has this issue. I absolutly have no idea what to do

**Comment by pulsarice** [comment]
I manually replaced tzdata in /data/misc/zoneinfo/ and /system/usr/share/zoneinfo/ and it worked on an old android 7.1 device. (Iran is +3:30 now)

**Comment by godverrdi** [comment]
I don't know what's wrong with my phone maybe it's reading time zones from some other file
Anyway thanks for your help :)

**Comment by kingofnull** [comment]
> I don't know what's wrong with my phone maybe it's reading time zones from some other file Anyway thanks for your help :)

How do you replace files from recovery or from android file manager?
If you are changing them from recovery and you have a stock miui ROM your changes will not apply to real booted system partition!

**Comment by theStarrk** [comment]
I replaced tzdata 2022g with tzdataN and tzdataO in module and installed it... and I checked it with file manager. my tzdata's size in /system/usr/share/zoneinfo/tzdata has changed so I can tell that it was replaced successfully... and also I'm not using miui
I didn't touch tz_version since I didn't know what to write in it
/data/misc/zoneinfo/ was empty still I added tzdata to /data/misc/zoneinfo/current to see if it works which it didn't

I tried TimeZone Fixer by force and it failed... but it also says that tzdata in my phone is 2022g
after that tried Timezone Updater which it failed too... and this is the log

[21:27:03] >>> TIME ZONE DATABASE
[21:27:03] Getting latest version ... wget: bad address 'data.iana.org'
[21:27:03] fail
[21:27:03] An error has occurred. Exiting.
[21:27:03] .
[21:27:03] >>> ICU DATA
[21:27:04] Getting latest version ... wget: bad address 'api.github.com'
[21:27:04] fail
[21:27:04] An error has occurred. Exiting.
[21:27:04] .

apparently it cannot reach the web address, I tried vpn but still failed.

now I'm stuck :/

**Comment by godverrdi** [comment]
> > I don't know what's wrong with my phone maybe it's reading time zones from some other file Anyway thanks for your help :)
> 
> How do you replace files from recovery or from android file manager? If you are changing them from recovery and you have a stock miui ROM your changes will not apply to real booted system partition!

I replaced tzdata 2022g with tzdataN and tzdataO in module and installed it... and I checked it with file manager. my tzdata's size in /system/usr/share/zoneinfo/tzdata has changed so I can tell that it was replaced successfully... and also I'm not using miui
I didn't touch tz_version since I didn't know what to write in it
/data/misc/zoneinfo/ was empty still I added tzdata to /data/misc/zoneinfo/current to see if it works which it didn't

I tried TimeZone Fixer by force and it failed... but it also says that tzdata in my phone is 2022g
after that tried Timezone Updater which it failed too... and this is the log

[21:27:03] >>> TIME ZONE DATABASE
[21:27:03] Getting latest version ... wget: bad address 'data.iana.org'
[21:27:03] fail
[21:27:03] An error has occurred. Exiting.
[21:27:03] .
[21:27:03] >>> ICU DATA
[21:27:04] Getting latest version ... wget: bad address 'api.github.com'
[21:27:04] fail
[21:27:04] An error has occurred. Exiting.
[21:27:04] .

apparently it cannot reach the web address, I tried vpn but still failed.

now I'm stuck :/

**Comment by kingofnull** [comment]
You must use tzdataN and rename it to `tzdata`. In some roms like MIUI you must root your device to force UI to read changes. Try magisk and report the result. A good way to investigate is to rename tzdata if your ui break after boot , you found the right file.

**Comment by waziig** [comment]
not working with miui android 10
i can confirm files in  /system/usr/share/zoneinfo/ are changed but nothing happened in time zones
i delete all files in zoneinfo and after reboot all back again so changing those file are useless

**Comment by VahidN** [comment]
These files should be updated:
    /data/misc/zoneinfo/tzdata or /system/usr/share/zoneinfo/*
    /system/usr/icu/*.dat
Source: https://github.com/meefik/tzupdater

**Comment by kingofnull** [comment]
> 



> These files should be updated: /data/misc/zoneinfo/tzdata or /system/usr/share/zoneinfo/* /system/usr/icu/*.dat Source: https://github.com/meefik/tzupdater

I almost sure /system/usr/icu/*.dat are only apples to Andriod 4 !

**Comment by kingofnull** [comment]
> not working with miui android 10 i can confirm files in /system/usr/share/zoneinfo/ are changed but nothing happened in time zones i delete all files in zoneinfo and after reboot all back again so changing those file are useless

Android 10 has implemented and zero free space portion lock on system partition you must resize partition to have some free space on system partition.
Resizing portions in android is dangerous and advance act. Back up all partitions specially presist partition.
You can read more about procedure here:
https://forum.xda-developers.com/t/guide-how-to-make-your-system-rw-super-r-w-read-write-able.4258745/

**Comment by kingofnull** [comment]
# ⚠️ Important Update ⚠️

###  `tzdata2021a` and lower doesn't cover Iran's new time zone! [❌](https://emojipedia.org/cross-mark/)

### Only `tzdata2022g` and newer is approved to fix Iran's new time zone. [✔️](https://emojipedia.org/check-mark/)

**Comment by waziig** [comment]
> > 
> 
> > These files should be updated: /data/misc/zoneinfo/tzdata or /system/usr/share/zoneinfo/* /system/usr/icu/*.dat Source: https://github.com/meefik/tzupdater
> 
> I almost sure /system/usr/icu/*.dat are only apples to Andriod 4 !

i have that (android10) and it is over 20MB in size so i didn't touch it
i used 2022g and 2023a tzdada and none of them works

**Comment by kingofnull** [comment]
## tzdate2022g replacement files:

- Android 9 Pie: [tzdata-2022g-android-pie-9.zip](https://github.com/Magisk-Modules-Repo/Systemless_TZData/files/11102300/tzdata-2022g-android-pie-9.zip)

- Android 10 Q: [tzdata-2022g-android-Q-10.zip](https://github.com/Magisk-Modules-Repo/Systemless_TZData/files/11102414/tzdata-2022g-android-Q-10.zip)

- Android 11 :  [tzdata-2022g-android-11.zip](https://github.com/Magisk-Modules-Repo/Systemless_TZData/files/11102558/tzdata-2022g-android-11.zip)

References: 

https://android.googlesource.com/platform/system/timezone/+/refs/heads/pie-dev

https://android.googlesource.com/platform/system/timezone/+/refs/heads/android10-dev

https://android.googlesource.com/platform/system/timezone/+/refs/heads/android11-dev

**Comment by godverrdi** [comment]
> > > 
> > 
> > 
> > > These files should be updated: /data/misc/zoneinfo/tzdata or /system/usr/share/zoneinfo/* /system/usr/icu/*.dat Source: https://github.com/meefik/tzupdater
> > 
> > 
> > I almost sure /system/usr/icu/*.dat are only apples to Andriod 4 !
> 
> i have that (android10) and it is over 20MB in size so i didn't touch it i used 2022g and 2023a tzdada and none of them works

Is there anyway to update this .dat file manually... for some reason it cannot download that file

**Comment by waziig** [comment]
> > > > 
> > > 
> > > 
> > > > These files should be updated: /data/misc/zoneinfo/tzdata or /system/usr/share/zoneinfo/* /system/usr/icu/*.dat Source: https://github.com/meefik/tzupdater
> > > 
> > > 
> > > I almost sure /system/usr/icu/*.dat are only apples to Andriod 4 !
> > 
> > 
> > i have that (android10) and it is over 20MB in size so i didn't touch it i used 2022g and 2023a tzdada and none of them works
> 
> Is there anyway to update this .dat file manually... for some reason it cannot download that file

if you referring to Timezone Updater app it wont work anymore, you need modded version that i cant find working download link
you can find more information here https://4pda.to/forum/index.php?showtopic=692805&st=160

**Comment by godverrdi** [comment]
> > > > > 
> > > > 
> > > > 
> > > > > These files should be updated: /data/misc/zoneinfo/tzdata or /system/usr/share/zoneinfo/* /system/usr/icu/*.dat Source: https://github.com/meefik/tzupdater
> > > > 
> > > > 
> > > > I almost sure /system/usr/icu/*.dat are only apples to Andriod 4 !
> > > 
> > > 
> > > i have that (android10) and it is over 20MB in size so i didn't touch it i used 2022g and 2023a tzdada and none of them works
> > 
> > 
> > Is there anyway to update this .dat file manually... for some reason it cannot download that file
> 
> if you referring to Timezone Updater app it wont work anymore, you need modded version that i cant find working download link you can find more information here https://4pda.to/forum/index.php?showtopic=692805&st=160

thanks a million

on 10+ one more path was added:
/apex/com.android.tzdata/etc/tz/tzdata

I just updated that file and boom... got rid of DST

**Comment by waziig** [comment]
> > > > > > 
> > > > > 
> > > > > 
> > > > > > These files should be updated: /data/misc/zoneinfo/tzdata or /system/usr/share/zoneinfo/* /system/usr/icu/*.dat Source: https://github.com/meefik/tzupdater
> > > > > 
> > > > > 
> > > > > I almost sure /system/usr/icu/*.dat are only apples to Andriod 4 !
> > > > 
> > > > 
> > > > i have that (android10) and it is over 20MB in size so i didn't touch it i used 2022g and 2023a tzdada and none of them works
> > > 
> > > 
> > > Is there anyway to update this .dat file manually... for some reason it cannot download that file
> > 
> > 
> > if you referring to Timezone Updater app it wont work anymore, you need modded version that i cant find working download link you can find more information here https://4pda.to/forum/index.php?showtopic=692805&st=160
> 
> thanks a million
> 
> on 10+ one more path was added: /apex/com.android.tzdata/etc/tz/tzdata
> 
> I just updated that file and boom... got rid of DST

THANK YOU! finally it worked
my phone didnt let me to change those files so i edited the customize.sh with new file addresses (/system/apex/com.android.tzdata/etc/tz/) and after enabling throw magisk it worked

**Comment by shiraziamir** [comment]
@waziig  @VahidN 

Would you kindly share your updated working module please.

**Comment by shiraziamir** [comment]
> > > > > > > 
> > > > > > 
> > > > > > 
> > > > > > > These files should be updated: /data/misc/zoneinfo/tzdata or /system/usr/share/zoneinfo/* /system/usr/icu/*.dat Source: https://github.com/meefik/tzupdater
> > > > > > 
> > > > > > 
> > > > > > I almost sure /system/usr/icu/*.dat are only apples to Andriod 4 !
> > > > > 
> > > > > 
> > > > > i have that (android10) and it is over 20MB in size so i didn't touch it i used 2022g and 2023a tzdada and none of them works
> > > > 
> > > > 
> > > > Is there anyway to update this .dat file manually... for some reason it cannot download that file
> > > 
> > > 
> > > if you referring to Timezone Updater app it wont work anymore, you need modded version that i cant find working download link you can find more information here https://4pda.to/forum/index.php?showtopic=692805&st=160
> > 
> > 
> > thanks a million
> > on 10+ one more path was added: /apex/com.android.tzdata/etc/tz/tzdata
> > I just updated that file and boom... got rid of DST
> 
> THANK YOU! finally it worked my phone didnt let me to change those files so i edited the customize.sh with new file addresses (/system/apex/com.android.tzdata/etc/tz/) and after enabling throw magisk it worked



> > > > > > > 
> > > > > > 
> > > > > > 
> > > > > > > These files should be updated: /data/misc/zoneinfo/tzdata or /system/usr/share/zoneinfo/* /system/usr/icu/*.dat Source: https://github.com/meefik/tzupdater
> > > > > > 
> > > > > > 
> > > > > > I almost sure /system/usr/icu/*.dat are only apples to Andriod 4 !
> > > > > 
> > > > > 
> > > > > i have that (android10) and it is over 20MB in size so i didn't touch it i used 2022g and 2023a tzdada and none of them works
> > > > 
> > > > 
> > > > Is there anyway to update this .dat file manually... for some reason it cannot download that file
> > > 
> > > 
> > > if you referring to Timezone Updater app it wont work anymore, you need modded version that i cant find working download link you can find more information here https://4pda.to/forum/index.php?showtopic=692805&st=160
> > 
> > 
> > thanks a million
> > on 10+ one more path was added: /apex/com.android.tzdata/etc/tz/tzdata
> > I just updated that file and boom... got rid of DST
> 
> THANK YOU! finally it worked my phone didnt let me to change those files so i edited the customize.sh with new file addresses (/system/apex/com.android.tzdata/etc/tz/) and after enabling throw magisk it worked



@waziig @VahidN

Would you kindly share your updated working module please and also how exactly should be done. I had no lock on android 12 and 13 while tried all the tips here.

**Comment by VahidN** [comment]
@shiraziamir 
[Follow these instructions in Persian.](https://www.dntips.ir/post/3283/%d8%b1%d9%88%d8%b4-%d9%86%d8%b5%d8%a8-%d8%a7%d9%86%d8%af%d8%b1%d9%88%db%8c%d8%af-10-%d8%a8%d8%b1-%d8%b1%d9%88%db%8c-%da%af%d9%88%d8%b4%db%8c%e2%80%8c%d9%87%d8%a7%db%8c-%d9%82%d8%af%db%8c%d9%85%db%8c-%d8%b3%d8%a7%d9%85%d8%b3%d9%88%d9%86%da%af#comment-19191)

**Comment by shiraziamir** [comment]
> @shiraziamir [Follow these instructions in Persian.](https://www.dntips.ir/post/3283/%d8%b1%d9%88%d8%b4-%d9%86%d8%b5%d8%a8-%d8%a7%d9%86%d8%af%d8%b1%d9%88%db%8c%d8%af-10-%d8%a8%d8%b1-%d8%b1%d9%88%db%8c-%da%af%d9%88%d8%b4%db%8c%e2%80%8c%d9%87%d8%a7%db%8c-%d9%82%d8%af%db%8c%d9%85%db%8c-%d8%b3%d8%a7%d9%85%d8%b3%d9%88%d9%86%da%af#comment-19191)

Thanks @VahidN
The key was to use Total Commander to replace the files in apex folder. I was successfull with Elixir custom rom (android 13) with just rooting and use TC.

**Comment by kingofnull** [comment]
> @shiraziamir [Follow these instructions in Persian.](https://www.dntips.ir/post/3283/%d8%b1%d9%88%d8%b4-%d9%86%d8%b5%d8%a8-%d8%a7%d9%86%d8%af%d8%b1%d9%88%db%8c%d8%af-10-%d8%a8%d8%b1-%d8%b1%d9%88%db%8c-%da%af%d9%88%d8%b4%db%8c%e2%80%8c%d9%87%d8%a7%db%8c-%d9%82%d8%af%db%8c%d9%85%db%8c-%d8%b3%d8%a7%d9%85%d8%b3%d9%88%d9%86%da%af#comment-19191)

Republish it on gist.github.com or github pages there tend to be more stable and available in feature.

**Comment by kingofnull** [comment]
> > > > > > > > 
> > > > > > > 
> > > > > > > 
> > > > > > > > These files should be updated: /data/misc/zoneinfo/tzdata or /system/usr/share/zoneinfo/* /system/usr/icu/*.dat Source: https://github.com/meefik/tzupdater
> > > > > > > 
> > > > > > > 
> > > > > > > I almost sure /system/usr/icu/*.dat are only apples to Andriod 4 !
> > > > > > 
> > > > > > 
> > > > > > i have that (android10) and it is over 20MB in size so i didn't touch it i used 2022g and 2023a tzdada and none of them works
> > > > > 
> > > > > 
> > > > > Is there anyway to update this .dat file manually... for some reason it cannot download that file
> > > > 
> > > > 
> > > > if you referring to Timezone Updater app it wont work anymore, you need modded version that i cant find working download link you can find more information here https://4pda.to/forum/index.php?showtopic=692805&st=160
> > > 
> > > 
> > > thanks a million
> > > on 10+ one more path was added: /apex/com.android.tzdata/etc/tz/tzdata
> > > I just updated that file and boom... got rid of DST
> > 
> > 
> > THANK YOU! finally it worked my phone didnt let me to change those files so i edited the customize.sh with new file addresses (/system/apex/com.android.tzdata/etc/tz/) and after enabling throw magisk it worked
> 
> > > > > > > > 
> > > > > > > 
> > > > > > > 
> > > > > > > > These files should be updated: /data/misc/zoneinfo/tzdata or /system/usr/share/zoneinfo/* /system/usr/icu/*.dat Source: https://github.com/meefik/tzupdater
> > > > > > > 
> > > > > > > 
> > > > > > > I almost sure /system/usr/icu/*.dat are only apples to Andriod 4 !
> > > > > > 
> > > > > > 
> > > > > > i have that (android10) and it is over 20MB in size so i didn't touch it i used 2022g and 2023a tzdada and none of them works
> > > > > 
> > > > > 
> > > > > Is there anyway to update this .dat file manually... for some reason it cannot download that file
> > > > 
> > > > 
> > > > if you referring to Timezone Updater app it wont work anymore, you need modded version that i cant find working download link you can find more information here https://4pda.to/forum/index.php?showtopic=692805&st=160
> > > 
> > > 
> > > thanks a million
> > > on 10+ one more path was added: /apex/com.android.tzdata/etc/tz/tzdata
> > > I just updated that file and boom... got rid of DST
> > 
> > 
> > THANK YOU! finally it worked my phone didnt let me to change those files so i edited the customize.sh with new file addresses (/system/apex/com.android.tzdata/etc/tz/) and after enabling throw magisk it worked
> 
> @waziig @VahidN
> 
> Would you kindly share your updated working module please and also how exactly should be done. I had no lock on android 12 and 13 while tried all the tips here.

I fix module problems with Magisk25 and add new files of 2022g and some testes it just safely replace a virtually mount directory and copy decent tzdata file to directory by android version.

Make sure reboot after installing module :

https://github.com/kingofnull/Systemless_TZData_2022g/releases/tag/beta

**Comment by kingofnull** [comment]
> > > > > > > > > 
> > > > > > > > 
> > > > > > > > 
> > > > > > > > > These files should be updated: /data/misc/zoneinfo/tzdata or /system/usr/share/zoneinfo/* /system/usr/icu/*.dat Source: https://github.com/meefik/tzupdater
> > > > > > > > 
> > > > > > > > 
> > > > > > > > I almost sure /system/usr/icu/*.dat are only apples to Andriod 4 !
> > > > > > > 
> > > > > > > 
> > > > > > > i have that (android10) and it is over 20MB in size so i didn't touch it i used 2022g and 2023a tzdada and none of them works
> > > > > > 
> > > > > > 
> > > > > > Is there anyway to update this .dat file manually... for some reason it cannot download that file
> > > > > 
> > > > > 
> > > > > if you referring to Timezone Updater app it wont work anymore, you need modded version that i cant find working download link you can find more information here https://4pda.to/forum/index.php?showtopic=692805&st=160
> > > > 
> > > > 
> > > > thanks a million
> > > > on 10+ one more path was added: /apex/com.android.tzdata/etc/tz/tzdata
> > > > I just updated that file and boom... got rid of DST
> > > 
> > > 
> > > THANK YOU! finally it worked my phone didnt let me to change those files so i edited the customize.sh with new file addresses (/system/apex/com.android.tzdata/etc/tz/) and after enabling throw magisk it worked
> > 
> > 
> > > > > > > > > 
> > > > > > > > 
> > > > > > > > 
> > > > > > > > > These files should be updated: /data/misc/zoneinfo/tzdata or /system/usr/share/zoneinfo/* /system/usr/icu/*.dat Source: https://github.com/meefik/tzupdater
> > > > > > > > 
> > > > > > > > 
> > > > > > > > I almost sure /system/usr/icu/*.dat are only apples to Andriod 4 !
> > > > > > > 
> > > > > > > 
> > > > > > > i have that (android10) and it is over 20MB in size so i didn't touch it i used 2022g and 2023a tzdada and none of them works
> > > > > > 
> > > > > > 
> > > > > > Is there anyway to update this .dat file manually... for some reason it cannot download that file
> > > > > 
> > > > > 
> > > > > if you referring to Timezone Updater app it wont work anymore, you need modded version that i cant find working download link you can find more information here https://4pda.to/forum/index.php?showtopic=692805&st=160
> > > > 
> > > > 
> > > > thanks a million
> > > > on 10+ one more path was added: /apex/com.android.tzdata/etc/tz/tzdata
> > > > I just updated that file and boom... got rid of DST
> > > 
> > > 
> > > THANK YOU! finally it worked my phone didnt let me to change those files so i edited the customize.sh with new file addresses (/system/apex/com.android.tzdata/etc/tz/) and after enabling throw magisk it worked
> > 
> > 
> > @waziig @VahidN
> > Would you kindly share your updated working module please and also how exactly should be done. I had no lock on android 12 and 13 while tried all the tips here.
> 
> I fix module problems with Magisk25 and add new files of 2022g and some testes it just safely replace a virtually mount directory and copy decent tzdata file to directory by android version.
> 
> Make sure reboot after installing module :
> 
> https://github.com/kingofnull/Systemless_TZData_2022g/releases/tag/beta

Try the module and report if there is bug or issue. This module is completely safe to install and it does not replace main system file it just make some directory overlays throw Magisk and can be remove without any footprints.

**Comment by kingofnull** [comment]
For those who cant find current tzdata in file system run a terminal with root access and run below  command to find real path of current tzdata : 

```bash
find / -name tzdata 2>/dev/null
```

**Comment by kingofnull** [comment]
More investigations shows in strange way apps that using time zone show clock incorrectly even after applying tzdata patch and while system ui shows time and timezone correctly! 

https://play.google.com/store/apps/details?id=nz.co.impressioncreative.timezone_viewer&hl=en&gl=US

**Comment by VahidN** [comment]
> More investigations shows in strange way apps that using time zone show clock incorrectly even after applying tzdata patch and while system ui shows time and timezone correctly!
> 
> https://play.google.com/store/apps/details?id=nz.co.impressioncreative.timezone_viewer&hl=en&gl=US

It says "Added missing "Myanmar" time zone" @Oct 31, 2020, which means it's using its own internal database for that and it's not using the system's database.

**Comment by kingofnull** [comment]
> > More investigations shows in strange way apps that using time zone show clock incorrectly even after applying tzdata patch and while system ui shows time and timezone correctly!
> > https://play.google.com/store/apps/details?id=nz.co.impressioncreative.timezone_viewer&hl=en&gl=US
> 
> It says "Added missing "Myanmar" time zone" @oct 31, 2020, which means it's using its own internal database for that and it's not using the system's database.

I don't think so In my phone , TimeZoneConverter app shows +4:30 timeozone as system timezone (the obsolete one). And some other apps have problems too like BCR: https://github.com/chenxiaolong/BCR

**Comment by kingofnull** [comment]
> > > More investigations shows in strange way apps that using time zone show clock incorrectly even after applying tzdata patch and while system ui shows time and timezone correctly!
> > > https://play.google.com/store/apps/details?id=nz.co.impressioncreative.timezone_viewer&hl=en&gl=US
> > 
> > 
> > It says "Added missing "Myanmar" time zone" @oct 31, 2020, which means it's using its own internal database for that and it's not using the system's database.
> 
> I don't think so In my phone , TimeZoneConverter app shows +4:30 timeozone as system timezone (the obsolete one). And some other apps have problems too like BCR: https://github.com/chenxiaolong/BCR



> > More investigations shows in strange way apps that using time zone show clock incorrectly even after applying tzdata patch and while system ui shows time and timezone correctly!
> > https://play.google.com/store/apps/details?id=nz.co.impressioncreative.timezone_viewer&hl=en&gl=US
> 
> It says "Added missing "Myanmar" time zone" @oct 31, 2020, which means it's using its own internal database for that and it's not using the system's database.

It is getting more confusing and strange , I have another phone with android 12 and I've never rooted or patched tzdata on it and It doing ok and automatically updated to new timezone 3:30 but the TimeZoneConverter app showing wrong timezone on this phone too!

**Comment by rxdelta** [comment]
I build tzdata 2023c successfully, but I cannot transfer the file to xiamo raphael (MI 9T Pro), path /system/usr/share/timezone/tzdata is read-only, and I cannot use adb command in recovery mode. is there any way to update the file without <root>ing device?

**Comment by kingofnull** [comment]
> I build tzdata 2023c successfully, but I cannot transfer the file to xiamo raphael (MI 9T Pro), path /system/usr/share/timezone/tzdata is read-only, and I cannot use adb command in recovery mode. is there any way to update the file without ing device?

I appreciate it if you share the build instruction with us (specially `icudt60i.dat` files), 

In short term your answer is a absolut No, You must root your device and we only tested MIUI with Magisk.

About transferring new files without root access a custom recovery like OrangeFox is needed And you must unlock your bootloader to install it which means you will lose all your current data. Also MIUI don't read `/system` partition modification. They implemented a unknown method to freeze system files probably in bootimage and Android read them from somewhere else , root is necessary to disable this mechanism.

**Comment by shiraziamir** [comment]
> > I build tzdata 2023c successfully, but I cannot transfer the file to xiamo raphael (MI 9T Pro), path /system/usr/share/timezone/tzdata is read-only, and I cannot use adb command in recovery mode. is there any way to update the file without ing device?
> 
> I appreciate it if you share the build instruction with us (specially `icudt60i.dat` files),
> 
> In short term your answer is a absolut No, You must root your device and we only tested MIUI with Magisk.
> 
> About transferring new files without root access a custom recovery like OrangeFox is needed And you must unlock your bootloader to install it which means you will lose all your current data. Also MIUI don't read `/system` partition modification. They implemented a unknown method to freeze system files probably in bootimage and Android read them from somewhere else , root is necessary to disable this mechanism.

I asked ChatGPT for help on how to build icudt.dat file. I need version 70i on my phone. While I managed to fix the main timezone and clock is fine, I still have issue with my smartphone and calendar on my phone. It seems still we need to update some more files. I don't have the tools to build it, so I just share it, hope it helps!

This is the answer from chatgpt:

The icudt70i.dat file is a part of the International Components for Unicode (ICU) library, which provides timezone support and other internationalization features for applications. Building a new icudt70i.dat file involves building the ICU library and then copying the generated file to your Android device.

Here's how you can build a new icudt70i.dat file for Android:

Download and install required tools:

- Install the latest version of Android NDK from https://developer.android.com/ndk/downloads
- Install Python 3.x and make sure it's added to your system's PATH variable.
- Download ICU source code:
  Download the latest ICU release from https://github.com/unicode-org/icu/releases
  Extract the contents of the downloaded file to a folder on your computer.

- Configure ICU for Android:

Open a command prompt or terminal window and navigate to the folder containing the extracted ICU source code.

Run the following command to configure ICU for Android, replacing <path_to_android_ndk> with the actual path to your Android NDK installation:

```
python icu4c/source/configure.py --with-cross-build=icu4c/source/data/in/icu-config --with-data-packaging=archive --disable-icu-config --disable-renaming --disable-layoutex --disable-extras --disable-tests --disable-samples --disable-strict --enable-trace --enable-static --host=arm-linux-androideabi --prefix=<path_to_android_ndk>/sysroot/usr
```

- Build ICU:

Run the following command to build ICU:

```
make
```
Once the build process is complete, you should find the icudt70i.dat file in the icu4c/source/data/out folder (assuming you're building ICU version 70).

**Comment by kingofnull** [comment]
> > > I build tzdata 2023c successfully, but I cannot transfer the file to xiamo raphael (MI 9T Pro), path /system/usr/share/timezone/tzdata is read-only, and I cannot use adb command in recovery mode. is there any way to update the file without ing device?
> > 
> > 
> > I appreciate it if you share the build instruction with us (specially `icudt60i.dat` files),
> > In short term your answer is a absolut No, You must root your device and we only tested MIUI with Magisk.
> > About transferring new files without root access a custom recovery like OrangeFox is needed And you must unlock your bootloader to install it which means you will lose all your current data. Also MIUI don't read `/system` partition modification. They implemented a unknown method to freeze system files probably in bootimage and Android read them from somewhere else , root is necessary to disable this mechanism.
> 
> I asked ChatGPT for help on how to build icudt.dat file. I need version 70i on my phone. While I managed to fix the main timezone and clock is fine, I still have issue with my smartphone and calendar on my phone. It seems still we need to update some more files. I don't have the tools to build it, so I just share it, hope it helps!
> 
> This is the answer from chatgpt:
> 
> The icudt70i.dat file is a part of the International Components for Unicode (ICU) library, which provides timezone support and other internationalization features for applications. Building a new icudt70i.dat file involves building the ICU library and then copying the generated file to your Android device.
> 
> Here's how you can build a new icudt70i.dat file for Android:
> 
> Download and install required tools:
> 
> * Install the latest version of Android NDK from https://developer.android.com/ndk/downloads
> * Install Python 3.x and make sure it's added to your system's PATH variable.
> * Download ICU source code:
>   Download the latest ICU release from https://github.com/unicode-org/icu/releases
>   Extract the contents of the downloaded file to a folder on your computer.
> * Configure ICU for Android:
> 
> Open a command prompt or terminal window and navigate to the folder containing the extracted ICU source code.
> 
> Run the following command to configure ICU for Android, replacing <path_to_android_ndk> with the actual path to your Android NDK installation:
> 
> ```
> python icu4c/source/configure.py --with-cross-build=icu4c/source/data/in/icu-config --with-data-packaging=archive --disable-icu-config --disable-renaming --disable-layoutex --disable-extras --disable-tests --disable-samples --disable-strict --enable-trace --enable-static --host=arm-linux-androideabi --prefix=<path_to_android_ndk>/sysroot/usr
> ```
> 
> * Build ICU:
> 
> Run the following command to build ICU:
> 
> ```
> make
> ```
> 
> Once the build process is complete, you should find the icudt70i.dat file in the icu4c/source/data/out folder (assuming you're building ICU version 70).

Seams promising !  But from things I red before it is much more complicated than it , rebuilding needs exact Android sources and It's ROM dependent. Actually some ICU builds are already available in [tzdata](https://android.googlesource.com/platform/system/timezone/+archive/refs/heads/android10-dev.tar.gz) (in output_data/icu_overlay) builds (I think those are what ChatGPT means) but they are not what we need.

Anyway if you succeed to build proper ICU files for your device please share the instructions.

**Comment by thisisamirmaleki** [comment]
I can't mount /apex/com.android.tzdata/... as rw...
This is the error in Total Commander:
`Re-mount error! mount: '/dev/block/loop8'->'/apex/com.android.tzdata': Invalid Argument`

**Comment by alirezaasol** [comment]
> > > > > > > > 
> > > > > > > 
> > > > > > > 
> > > > > > > > These files should be updated: /data/misc/zoneinfo/tzdata or /system/usr/share/zoneinfo/* /system/usr/icu/*.dat Source: https://github.com/meefik/tzupdater
> > > > > > > 
> > > > > > > 
> > > > > > > I almost sure /system/usr/icu/*.dat are only apples to Andriod 4 !
> > > > > > 
> > > > > > 
> > > > > > i have that (android10) and it is over 20MB in size so i didn't touch it i used 2022g and 2023a tzdada and none of them works
> > > > > 
> > > > > 
> > > > > Is there anyway to update this .dat file manually... for some reason it cannot download that file
> > > > 
> > > > 
> > > > if you referring to Timezone Updater app it wont work anymore, you need modded version that i cant find working download link you can find more information here https://4pda.to/forum/index.php?showtopic=692805&st=160
> > > 
> > > 
> > > thanks a million
> > > on 10+ one more path was added: /apex/com.android.tzdata/etc/tz/tzdata
> > > I just updated that file and boom... got rid of DST
> > 
> > 
> > THANK YOU! finally it worked my phone didnt let me to change those files so i edited the customize.sh with new file addresses (/system/apex/com.android.tzdata/etc/tz/) and after enabling throw magisk it worked
> 
> > > > > > > > 
> > > > > > > 
> > > > > > > 
> > > > > > > > These files should be updated: /data/misc/zoneinfo/tzdata or /system/usr/share/zoneinfo/* /system/usr/icu/*.dat Source: https://github.com/meefik/tzupdater
> > > > > > > 
> > > > > > > 
> > > > > > > I almost sure /system/usr/icu/*.dat are only apples to Andriod 4 !
> > > > > > 
> > > > > > 
> > > > > > i have that (android10) and it is over 20MB in size so i didn't touch it i used 2022g and 2023a tzdada and none of them works
> > > > > 
> > > > > 
> > > > > Is there anyway to update this .dat file manually... for some reason it cannot download that file
> > > > 
> > > > 
> > > > if you referring to Timezone Updater app it wont work anymore, you need modded version that i cant find working download link you can find more information here https://4pda.to/forum/index.php?showtopic=692805&st=160
> > > 
> > > 
> > > thanks a million
> > > on 10+ one more path was added: /apex/com.android.tzdata/etc/tz/tzdata
> > > I just updated that file and boom... got rid of DST
> > 
> > 
> > THANK YOU! finally it worked my phone didnt let me to change those files so i edited the customize.sh with new file addresses (/system/apex/com.android.tzdata/etc/tz/) and after enabling throw magisk it worked
> 
> @waziig @VahidN
> 
> Would you kindly share your updated working module please and also how exactly should be done. I had no lock on android 12 and 13 while tried all the tips here.

i thought that android ver 10 and above updates tzdata independently and there is no need  to update firmware.

**Comment by marc0601** [comment]
> > > > > > 
> > > > > 
> > > > > 
> > > > > > These files should be updated: /data/misc/zoneinfo/tzdata or /system/usr/share/zoneinfo/* /system/usr/icu/*.dat Source: https://github.com/meefik/tzupdater
> > > > > 
> > > > > 
> > > > > I almost sure /system/usr/icu/*.dat are only apples to Andriod 4 !
> > > > 
> > > > 
> > > > i have that (android10) and it is over 20MB in size so i didn't touch it i used 2022g and 2023a tzdada and none of them works
> > > 
> > > 
> > > Is there anyway to update this .dat file manually... for some reason it cannot download that file
> > 
> > 
> > if you referring to Timezone Updater app it wont work anymore, you need modded version that i cant find working download link you can find more information here https://4pda.to/forum/index.php?showtopic=692805&st=160
> 
> thanks a million
> 
> on 10+ one more path was added: /apex/com.android.tzdata/etc/tz/tzdata
> 
> I just updated that file and boom... got rid of DST


@theStarrk hey can you provide the flashable file that you used to get it work or the way you replaced the files in /apex/com.android.tzdata/etc/tz/tzdata , because i tried manually it can't be edited and tried to modify the magisk module and still no modification to this dir after flashing

@waziig i also added /system/apex/com.android.tzdata/etc/tz/ this address to the module and the files here are okay but still the only dir that contain stock files is /apex is there anyway that i can modify it??

**Comment by godverrdi** [comment]
> @theStarrk hey can you provide the flashable file that you used to get it work or the way you replaced the files in /apex/com.android.tzdata/etc/tz/tzdata , because i tried manually it can't be edited and tried to modify the magisk module and still no modification to this dir after flashing
> 
> @waziig i also added /system/apex/com.android.tzdata/etc/tz/ this address to the module and the files here are okay but still the only dir that contain stock files is /apex is there anyway that i can modify it??

I don't have the flashable file... I replaced tzdata with file manager(mix)
But you should know to overwrite a file in root, you can't paste it directly to the path... first paste it to another directory that doesn't have a file with similar name, then move it to /apex/com.android.tzdata/etc/tz/

**Comment by marc0601** [comment]
> > @theStarrk hey can you provide the flashable file that you used to get it work or the way you replaced the files in /apex/com.android.tzdata/etc/tz/tzdata , because i tried manually it can't be edited and tried to modify the magisk module and still no modification to this dir after flashing
> > @waziig i also added /system/apex/com.android.tzdata/etc/tz/ this address to the module and the files here are okay but still the only dir that contain stock files is /apex is there anyway that i can modify it??
> 
> I don't have the flashable file... I replaced tzdata with file manager(mix) But you should know to overwrite a file in root, you can't paste it directly to the path... first paste it to another directory that doesn't have a file with similar name, then move it to /apex/com.android.tzdata/etc/tz/


I am also using mixplorer, but /Apex dir is Read only i can't replace the original file in /apex/com.android.tzdata/etc/tz/, I'm using android 11 miui eu 12.5 with magisk 26.1

**Comment by godverrdi** [comment]
> I am also using mixplorer, but /Apex dir is Read only i can't replace the original file in /apex/com.android.tzdata/etc/tz/, I'm using android 11 miui eu 12.5 with magisk 26.1

Just to make sure... Check Magisk to see if mixplorer has premission to use root
I'm not using miui but i did this on a miui device so I know it can be done

**Comment by be4zad** [comment]
In my Samsung phone with Android 11, I'm just replaced tzdata and tz_version to `/system/apex/com.android.tzdata/etc/tz/` and it's worked!

**Comment by Navid14101** [comment]
It took my time to find out so I'm sharing this here.
If you have Android 12 due to the changes Google made you can't access system files easily.
So you have to follow this guide for replacing "tz_version" & "tzdata" in the following path:
https://forum.xda-developers.com/t/how-to-change-any-file-or-directory-using-magisk.4543103/

**Comment by thisisamirmaleki** [comment]
> It took my time to find out so I'm sharing this here. If you have Android 12 due to the changes Google made you can't access system files easily. So you have to follow this guide for replacing "tz_version" & "tzdata" in the following path: https://forum.xda-developers.com/t/how-to-change-any-file-or-directory-using-magisk.4543103/

Can you please make a tutorial or share your script?

**Comment by Xosrov** [comment]
@thisisamirmaleki 
I wrote a Magisk "module" for this. Since newer androids have read-only file system, Magisk is required for persistent changes.
check [this](https://github.com/Xosrov/Android-Timezone-Fix) out.

Thanks to @Navid14101 for the guide he shared

**Comment by thisisamirmaleki** [comment]
> @thisisamirmaleki I wrote a Magisk "module" for this. Since newer androids have read-only file system, Magisk is required for persistent changes. check [this](https://github.com/Xosrov/Android-Timezone-Fix) out.
> 
> Thanks to @Navid14101 for the guide he shared

I'm using Arrow OS 11 and tried this without any luck, still on +4:30 timezone...

**Comment by Xosrov** [comment]
> > @thisisamirmaleki I wrote a Magisk "module" for this. Since newer androids have read-only file system, Magisk is required for persistent changes. check [this](https://github.com/Xosrov/Android-Timezone-Fix) out.
> > Thanks to @Navid14101 for the guide he shared
> 
> I'm using Arrow OS 11 and tried this without any luck, still on +4:30 timezone...

1. `adb shell` into running device and go root with `su -` (Magisk will ask for permission).
2. `cd /`
3. `find . -name tzdata 2>/dev/nul`

Create an issue in the [repository](https://github.com/Xosrov/Android-Timezone-Fix) and attach the output of the last command. I'll see what I can do. We can continue this discussion there.

**Comment by be4zad** [comment]
> > @thisisamirmaleki I wrote a Magisk "module" for this. Since newer androids have read-only file system, Magisk is required for persistent changes. check [this](https://github.com/Xosrov/Android-Timezone-Fix) out.
> > Thanks to @Navid14101 for the guide he shared
> 
> I'm using Arrow OS 11 and tried this without any luck, still on +4:30 timezone...

Please also try this module and see if it works or not
[Android-Timezone-Fix-v1-system_ext.zip](https://github.com/Magisk-Modules-Repo/Systemless_TZData/files/11621685/Android-Timezone-Fix-v1-system_ext.zip)

**Comment by thisisamirmaleki** [comment]
> > > @thisisamirmaleki I wrote a Magisk "module" for this. Since newer androids have read-only file system, Magisk is required for persistent changes. check [this](https://github.com/Xosrov/Android-Timezone-Fix) out.
> > > Thanks to @Navid14101 for the guide he shared
> > 
> > 
> > I'm using Arrow OS 11 and tried this without any luck, still on +4:30 timezone...
> 
>     1. `adb shell` into running device and go root with `su -` (Magisk will ask for permission).
> 
>     2. `cd /`
> 
>     3. `find . -name tzdata 2>/dev/nul`
> 
> 
> Create an issue in the [repository](https://github.com/Xosrov/Android-Timezone-Fix) and attach the output of the last command. I'll see what I can do. We can continue this discussion there.

I updated my OS to ArrowOS 13.1 and I CAN mount /system/apex ... as RW. However I also tested your Magisk Module thingy, and it works like a charm. Thank you.

**Comment by thisisamirmaleki** [comment]
@Xosrov I came across another issue while using your module. The time's alright, the DATE on the status bar however, is not correct from 23 to 24 everyday.
It might also be worth noting that the time widget is showing the wrong TIME too.

**Comment by KhMahallati** [comment]
> > > @thisisamirmaleki I wrote a Magisk "module" for this. Since newer androids have read-only file system, Magisk is required for persistent changes. check [this](https://github.com/Xosrov/Android-Timezone-Fix) out.
> > > Thanks to @Navid14101 for the guide he shared
> > 
> > 
> > I'm using Arrow OS 11 and tried this without any luck, still on +4:30 timezone...
> 
> Please also try this module and see if it works or not [Android-Timezone-Fix-v1-system_ext.zip](https://github.com/Magisk-Modules-Repo/Systemless_TZData/files/11621685/Android-Timezone-Fix-v1-system_ext.zip)

I use ArrowOS 13.1 on Xiaomi Note 7 (lavender) . I tested this module But no luck?
what can i do? plz help

**Comment by KhMahallati** [comment]
the module installed successfully via magisk. but the time zone is still +04:30 for Iran

**Comment by dev4min** [comment]
> [tzdata-2022g.zip](https://github.com/Magisk-Modules-Repo/Systemless_TZData/files/11051129/tzdata-2022g.zip)

It worked like a charm!
Now Tehran time zone is `+03:30`
Thank you :)
`LineageOS 14.1 (Android 7.1.2)`
